File: rpi/tests/uart/protocol.py
"""
protocol.py

Format:

COMMAND|DATA

Beispiele:

GET|TEXT

TEXT|Hallo Welt
"""


class Protocol:


    SEPARATOR = "|"



    @staticmethod
    def encode(
        command,
        data=""
    ):

        return (
            command
            +
            Protocol.SEPARATOR
            +
            data
            +
            "\n"
        )



    @staticmethod
    def decode(
        message
    ):

        parts = message.split(
            Protocol.SEPARATOR,
            1
        )


        command = parts[0]

        data = ""


        if len(parts) > 1:

            data = parts[1]


        return command, data

File: esp/lib/communication/protocol.cpp
#include "protocol.h"


String Protocol::encode(
    String command,
    String data
)
{

    return command +
           "|" +
           data +
           "\n";

}



bool Protocol::decode(
    String message,
    String &command,
    String &data
)
{

    int pos = message.indexOf(
        '|'
    );


    if(pos < 0)
    {
        return false;
    }


    command = message.substring(
        0,
        pos
    );


    data = message.substring(
        pos + 1
    );


    data.trim();


    return true;

}